resources/gcloud.rb

```diff
@@ -9,90 +9,284 @@
 end
 
 action :install do
-  platform_cmd = Mixlib::ShellOut.new('uname')
-  platform_cmd.run_command
-  platform_cmd.error!
-  platform = platform_cmd.stdout.strip.downcase
-
-  arch_cmd = Mixlib::ShellOut.new('uname -m')
-  arch_cmd.run_command
-  arch_cmd.error!
-  arch = arch_cmd.stdout.strip
-
-  version = new_resource.version
-  if version.empty?
+  # Check infomation of operation system
+  platform = attr_platform
+  arch = attr_arch
+  arch_core = attr_arch_core(arch)
+
+  # version will be installed
+  latest_version = attr_latest_version
+  version = if new_resource.version.empty?
+              latest_version
+            else
+              new_resource.version
+            end
+  existing_version = attr_existing_version
+
+  # Compare existing_version vs version
+  if existing_version != version
+    # Deleting previous version if mismatched
+    delete_gcloud(binary_path)
+
+    # Check version if version avaiable in apt-get package (ubuntu)
+    # Avaiable version get from https://packages.cloud.google.com/apt/dists/cloud-sdk-$('lsb_release -c -s')/main/binary-#{arch}/Packages
+    # Installing gcloud autocomplete via apt-get will be faster
+    # issue (https://github.com/teracyhq-incubator/kubernetes-stack-cookbook/issues/19)
+    if platform?('ubuntu')
+      version_avaiable = version_avaiable_in_apt_package(arch_core, version) unless arch_core.to_s.strip.empty?
+    end
+
+    # Gcloud version will be installed via apt-get
+    if version_avaiable == true
+      execute 'import google-cloud-sdk public key' do
+        command 'curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -'
+      end
+
+      apt_repository 'google-cloud-sdk' do
+        uri          'http://packages.cloud.google.com/apt'
+        distribution "cloud-sdk-#{node['lsb']['codename']}"
+        components   ['main']
+        # key 'A7317B0F'
+        # keyserver 'packages.cloud.google.com/apt/doc/apt-key.gpg'
+      end
+
+      if version == latest_version
+        package 'google-cloud-sdk'
+      else
+        package 'google-cloud-sdk' do
+          version "#{version}-0"
+        end
+      end
+
+      # Link to binary_path
+      link_to_binary_path(binary_path)
+      # Disable update notification when run command 'gcloud version'
+      disable_update_check
+    # Gcloud version will be installed via file downloaded
+    else
+      # Check requirement to install
+      install_requirement_for_ubuntu if platform?('ubuntu')
```

Comment:
I guess we can combine both: install_requirements() and the platform checking and handling should be done on that method instead of 2 calls. Later on, we support more platforms, add more checking like this is weird.